Making ops c10-full: optional lists

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49259 Remove `use_c10_dispatcher: full` lines
* #49251 Remove generated_unboxing_wrappers and setManuallyBoxedKernel
* #49220 Remove .impl_UNBOXED() and functionalities associated with it
* #49164 Remove codegen logic to support non-c10-full ops
* #49138 [wip] Making ops c10-full: list of optional tensors
* #49150 Making ops c10-full: ConstQuantizerPtr
* #49146 Making ops c10-full: Storage arguments
* **#49088 Making ops c10-full: optional lists**
* #49083 Making ops c10 full: optional out arguments

We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

smessmer committed:
Making ops c10-full: optional lists

We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

ezyang reviewed (approved):
Vector allocation here is still giving me hives. It looks like the operators affected here are only cold names operators though. @swolchok I cc'ed you specifically about the vector allocation in aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h

swolchok commented:
> Vector allocation here is still giving me hives. It looks like the operators affected here are only cold names operators though. @swolchok I cc'ed you specifically about the vector allocation in aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h

I don’t see a new allocation in that file, just a generalization of existing special cases. What am I missing?

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

ezyang commented:
> I don’t see a new allocation in that file, just a generalization of existing special cases. What am I missing?

I don't think this PR makes it worse. But it touches code that is not so great.

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: optional lists"


We had special case logic to support `int[]?` and `double[]?` but nothing for `DimnameList[]?`.
This PR generalizes the logic to support optional lists so it should now work with all types.
It also enables c10-fullness for ops that were blocked by this.

Note that using these arguments in a signature was always and still is expensive because the whole list needs to be copied.
We should probably consider alternatives in the future like for example using `torch::List` instead of `ArrayRef`, that could work without copying the list.

Differential Revision: [D25423901](https://our.internmc.facebook.com/intern/diff/D25423901/)

[ghstack-poisoned]
